### ISSUE TITLE
no_docstrings not working for stdlib

### ISSUE THREAD

**Issue by MMMZZZZ** [Nuitka/Nuitka#975]
### Short description: 

As the title says, docstrings are apparently not removed from the stdlib files even though `no_docstrings` is specified.

### Example

Source file: 
```
"""This docsting will be removed as expected. """

if __name__ == "__main__":
    print("Hello World!")
```
Nuitka Command Line: 
```
python -m nuitka docstring-test.py --standalone --onefile --python-flag=no_site,no_asserts,no_docstrings --windows-company-name Zedia --windows-product-version 4096
```

Result:
The resulting executable does not contain the docstring from the source file, but it does contain docstrings from the stdlib files. One specific example I found in the executable is `"Heap queue algorithm (a.k.a. priority queue)" [...]` - a docstring from the beginning of the file `[pythondir]/Lib/heapq.py`.

The issue remains if I add the flag `--follow-stdlib` (hoped the C compiler would maybe remove it). 

Haven't tested if this issue concerns other imports, too.

### Additional Info

> Nuitka version, full Python version and Platform (Windows, OSX, Linux ...)

Tested with both stable and current development versions - same issue.
```(venv) D:\Dokumente\PycharmProjects\Scratchbook>python -m nuitka --version
0.6.11.5
Python: 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)]
Executable: D:\Dokumente\PycharmProjects\Scratchbook\venv\Scripts\python.exe
OS: Windows
Arch: x86_64
```
```(venv) D:\Dokumente\PycharmProjects\Scratchbook>python -m nuitka --version
0.6.12rc5
Python: 3.9.0 (tags/v3.9.0:9cf6752, Oct  5 2020, 15:34:40) [MSC v.1927 64 bit (AMD64)]
Executable: D:\Dokumente\PycharmProjects\Scratchbook\venv\Scripts\python.exe
OS: Windows
Arch: x86_64
```

> How did you install Nuitka and Python

using pip within the venv of my PyCharms project. 

> The specific PyPI names and versions

Actually, I have quite a lot of packages installed in this venv. However in this specific example I'm not using anything except the stdlib so I'd argue the list is not important for this issue. Please correct me if I'm wrong.

**Comment by kayhayen** [comment]
That's actually a good catch, I doubt that docstrings are any use in there, although there might be code relying on it.

Our difficulty here, will be the following:

                elif part in ("-O", "no_asserts", "noasserts"):
                    _python_flags.add("no_asserts")
                elif part in ("no_docstrings", "nodocstrings"):
                    _python_flags.add("no_docstrings")
                elif part in ("-OO",):
                    _python_flags.add("no_docstrings")
                    _python_flags.add("no_asserts")

As you can see, `no_docstrings` is a Nuitka specialty, which doesn't exist in Python. And when we compile bytecode, we might not have the ability to do only one and not both, I will check it.

It's also noteworthy, that some OSes, have stopped including source code for some of standard library, Fedora 33 is something to consider there, in that case, we will only be able to do it, by patching the bytecode itself.

**Comment by kayhayen** [comment]
I checked the `compile` built-in that is currently used to turn source code to bytecode there, and it doesn't consider any of the flags.

What would have to the done, is to use `ast.parse`, which is going to give us a chance to modify the tree, e.g. setting all `doc` attributes to `None`. The passing that tree to `compile` is going to work nicely. Doing it on the resulting bytecode object from direct `compile` may make it more likely to work in all cases. The `ast.parse` is known to fail with recursion errors on too complex code.

**Comment by kayhayen** [comment]
Deleting the docstrings causes segfaults to Python, and setting them to None caused errors on future imports, so setting them to empty string is the safest way. Since that's a shared value too, it doesn't matter much I guess. Not doing it for demoted to bytecode modules now, not until I have a good test case that shows that we can do a fallback to primitive bytecode compilation, should that `ast.parse` fail.

**Comment by MMMZZZZ** [comment]
Thanks for the quick reaction!

I have to say, I can't follow you a 100% but then I'm not that experienced with Python (`ast` definitely is something I'd need a LOT more time to understand). 
However, I can tell you what my original goal was. When I compile the hello world example from above into a standalone exe I get a 24MB (!) file. That seemed crazy to me. Adding `--follow-stdlib` reduced it to about 7MB - much better but still not exactly small. So I had a look at the exe and found tons of (doc)strings. 
Since I can't judge how complicated it is to implement this (and hence how long it might take) I'd like to know if there is something I could do myself to get rid of that unecessary data in the exe? 

Kind regards,
Max

**Comment by kayhayen** [comment]
With the `--follow-stdlib` there should have no docstrings all along, can you confirm or deny that?

I have the change ready to be pushed on factory, but currently a pre-release is building and I don't want to mess with that yet, I will let you know.

With that, docstrings for standard library should be gone either way. As an aside, you can also try the no_site, which prevents including things that `site.py` loads (and never get used), which somtimes can be a lot too.

I am not sure if `--follow-stdlib` ought to reduce it that much, or rather increase it. But that's a clear hint that we need to have it on by default, even if we demoted it later again. But right now, before caching of result of Nuitka frontend compilation, I don't want to go there, it would just take too much of my time in this order.

**Comment by kayhayen** [comment]
Also, keep in mind, your program may say hello world only, but you asked to bring a Python installation along. These are not small.

**Comment by MMMZZZZ** [comment]
> With the `--follow-stdlib` there should have no docstrings all along, can you confirm or deny that?

That option *does* reduce the file size drastically but it *does not* remove the docstrings from the stdlib files as states in the initial post. 

I can try to look a little more into what causes the drastic size difference of `--follow-stdlib`. However, I don’t know if I’ll find anything meaningful since I’m not experienced at all with how Nuitka works and thus what I should expect or not. 

> I have the change ready to be pushed on factory

Awesome, thank you!

**Comment by kayhayen** [comment]
@MMMZZZZ can you provide evidence of the doc string not being reduced from the module or a function? I would need an example to work with. From what I saw, it did it. But the visitor might miss some things of course.

**Comment by MMMZZZZ** [comment]
I'm afraid I can't quite follow you (as I said, not very experienced in this field). What evidence or example do you need...? First post contains the command line, the source that's been compiled and an example of a docstring that has not been removed. All I can think of that's missing would be the .exe file(s) themselves?

Regards,
Max

**Comment by kayhayen** [comment]
I am missing the docstring that is included and you think it should be removed, with a pointer to the source of the standard library where it lives. If you see data in the binary, it doesn't have to be doc strings, it can be regular strings too. At this time I don't know your example of a docstring that survives.

**Comment by MMMZZZZ** [comment]
Quote from the opening post:

> Result:
> The resulting executable does not contain the docstring from the source file, but it does contain docstrings from the stdlib files. **One specific example I found in the executable is `"Heap queue algorithm (a.k.a. priority queue)" [...]` - a docstring from the beginning of the file `[pythondir]/Lib/heapq.py`.**
> 
> The issue remains if I add the flag `--follow-stdlib` (hoped the C compiler would maybe remove it). 

To rephrase it using quotes from your post: 

> the docstring that is included and you think it should be removed

`"Heap queue algorithm (a.k.a. priority queue)"` (the string is much longer but this snipped is enough to identify it).

> a pointer to the source of the standard library where it lives

docstring from the beginning of the file `[pythondir]/Lib/heapq.py`.

> If you see data in the binary, it doesn't have to be doc strings, it can be regular strings too. 

Of course. But I'm pretty sure the given example is a docstring. And it's not the only one; it just happened to be the first one I found. 

I can only assume you overlooked the given example. Otherwise I'm sorry but I'll need more explanations what information is missing in the example and how to provide a more useful one

Regards,
Max

**Comment by kayhayen** [comment]
We might be talking past each other. The change is part of the 0.6.12 release and previously was on factory, which I forgot to state, did you have it occur using that version, because it seems you think I dispute your original report, which I didn't mean to. I cannot reproduce your example string in either Python2 not Python3, so I am assuming we are just confused (and it's my fault definitely ?

**Comment by MMMZZZZ** [comment]
Oh okay. Yeah I really didn't realize you released a new versions that's supposed to fix it. Unfortunately I have bad news: the new version didn't fix it. Here's what I did:

 * Installed Nuitka v0.6.12.2
 * Deleted all the folders with cached files from previous runs.
 * Used the command line and example file I posted at the top
 * **Result: heap queue docstring and many others still present in the exe**
 * Added `--follow-stdlib`, compiled again
 * **Result: heap queue docstring and many others still present in the exe**

Aaaand the heap queue docstring is still contained in the exe. Others, too. 

Regards,
Max

**Comment by kayhayen** [comment]
Ok, I was using my development version, but I don't think it should make a difference, and could find it with either `vi` nor `strings` on the executable, i saw a lot of things that looked like docstrings though. I didn't check though if "heap q" which is what I looked for is actually in the source files, which I now did.

May I suggest you check if you get into `compileSourceToBytecode` of `nuitka.Bytecodes` module and see what it does, and report if it doesn't remove things for you or doesn't get asked? Maybe your installation has no source code for that standard library module, and you don't take that code path?
